psydac/linalg/solvers.py

```diff
@@ -1725,3 +1728,260 @@ def solve(self, b, out=None):
+        
+        # Extract local storage
+        r = self._tmps["r"]
+        v = self._tmps["v"]
+
+        # Internal objects of GMRES
+        H = self._H
+        Q = []
+        beta = []
+        sn = []
+        cn = []
+
+        # First values
+
+        A.dot( x , out=r)
+
+        r += b - r 
+
+        r_norm = r.dot(r).real ** 0.5
+        am = r_norm
+
+        beta.append(r_norm)
+        Q.append(r / r_norm)
+
+
+        if verbose:
+            print( "GMRES solver:" )
+            print( "+---------+---------------------+")
+            print( "+ Iter. # | L2-norm of residual |")
+            print( "+---------+---------------------+")
+            template = "| {:7d} | {:19.2e} |"
+            print( template.format( 1, am ) )
+
+        # Iterate to convergence
+
+        for k in range(maxiter):
+            if am < tol:
+                break
+
+            # run Arnoldi
+            self.arnoldi(k, Q)
+
+            # make the last diagonal entry in H equal to 0, so that H becomes upper triangular
+            self.apply_givens_rotation(k, sn, cn)
+
+            # update the residual vector
+            beta.append(- sn[k] * beta[k])
+            beta[k] *= cn[k]
+
+            am = abs(beta[k+1])
+            if verbose:
+                print( template.format( k+2, am ) )
+
+        if verbose:
+            print( "+---------+---------------------+")        
+        # calculate result
+        y = self.solve_triangular(H[:k, :k], beta[:k]) # system of upper triangular matrix
+
+        for i in range(k):
+            Q[i].copy(out=v)
+            v *= y[i]
+            x += v
+        
+        #x[x.starts[0] : x.ends[0] + 1] += Q[:m] @ y
+
+        # Convergence information
+        self._info = {'niter': k+1, 'success': am < tol, 'res_norm': am }
+        
+        return x
+    
+    def solve_triangular(self, T, d):
+        # Backwards substitution. Assumes T is upper triangular
+        k = T.shape[0]
+        y = np.zeros((k,), dtype=self._A.dtype)
+
+        for k1 in range(k):
+            temp = 0.
+            for k2 in range(1, k1 + 1):
+                temp += T[k - 1 - k1, k - 1 - k1 + k2] * y[k - 1 - k1 + k2]
+            y[k - 1 - k1] = ( d[k - 1 - k1] - temp ) / T[k - 1 - k1, k - 1 - k1]
+        
+        return y
+
+    def arnoldi(self, k, Q):
+        h = self._H[:k+2, k]
+
+        p = self._tmps["p"]
+        self._A.dot( Q[k] , out=p) # Krylov vector
+
+        for i in range(k + 1): # Modified Gram-Schmidt, keeping Hessenberg matrix
+            h[i] = p.dot( Q[i] )
+            p -= h[i] * Q[i]
+        
+        h[k+1] = p.dot(p) ** 0.5
+        p /= h[k+1] # Normalize vector
+
+        Q.append(p.copy())
+
+
+    def apply_givens_rotation(self, k, sn, cn):
+        # Apply Givens rotation to last column of H
+        h = self._H[:k+2, k]
+
+        for i in range(k):
+            temp = cn[i] * h[i] + sn[i] * h[i+1]
+            h[i+1] = - sn[i] * h[i] + cn[i] * h[i+1]
```

Comment:
Should optimize the code with only internal operation:
h[i+1] *= cn[i], h[i] *= sn[i], h[i+1] -= h[i] 

The same should be done for temp